src/modules/differential_drive_control/DifferentialDriveGuidance/DifferentialDriveGuidance.hpp

```diff
@@ -0,0 +1,140 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include <px4_platform_common/module.h>
+#include <px4_platform_common/module_params.h>
+
+#include <matrix/matrix/math.hpp>
+#include <lib/geo/geo.h>
+#include <math.h>
+
+#include <lib/motion_planning/PositionSmoothing.hpp>
+#include <lib/motion_planning/VelocitySmoothing.hpp>
+
+#include <lib/pid/pid.h>
+
+/**
+ * @brief Enum class for the different states of guidance.
+ */
+enum class GuidanceState {
+	TURNING, ///< The vehicle is currently turning.
+	DRIVING, ///< The vehicle is currently driving straight.
+	GOAL_REACHED ///< The vehicle has reached its goal.
+};
+
+/**
+ * @brief Class for differential drive guidance.
+ */
+class DifferentialDriveGuidance : public ModuleParams
+{
+public:
+	/**
+	 * @brief Constructor for DifferentialDriveGuidance.
+	 * @param parent The parent ModuleParams object.
+	 */
+	DifferentialDriveGuidance(ModuleParams *parent);
+
+	/**
+	 * @brief Default destructor.
+	 */
+	~DifferentialDriveGuidance() = default;
+
+	/**
+	 * @brief Compute guidance for the vehicle.
+	 * @param global_pos The global position of the vehicle.
+	 * @param current_waypoint The current waypoint the vehicle is heading towards.
+	 * @param next_waypoint The next waypoint the vehicle will head towards after reaching the current waypoint.
+	 * @param vehicle_yaw The yaw orientation of the vehicle.
+	 * @param body_velocity The velocity of the vehicle.
+	 * @param angular_velocity The angular velocity of the vehicle.
+	 * @param dt The time step.
+	 * @return A 2D vector containing the computed guidance.
```

Comment:
The speed setpoint, is the forward body velocity for the rover which is also the input into the inverse kinematics library. which then calculates the required wheel speeds, while respecting saturation. 

If I create an enum in the guidance controller, I would also need to add it into the DifferentialDriveControl.cpp file, would you recommend that? I am not yet sure what the best balance of clarity / more code is. I improved the @brief, maybe this is enough? 

`	 * @return A 2D vector containing the computed guidance (speed in m/s, yaw rate in rad/s).
`

If not please let me know, and I will happily implement the enum!

Reply:
Aren't they inheriting from each other? Then you'd only need to implement it at one place no?
I still think it would be cleaner, but am curious for other inputs, eg @MaEtUgR ?